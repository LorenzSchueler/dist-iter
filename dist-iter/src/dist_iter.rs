use mpi::{
    topology::SimpleCommunicator,
    traits::{Communicator, Source},
};

use crate::task::Task;

pub struct DistIter<'w, I>
where
    I: Iterator,
    I::Item: Task,
{
    inner: I,
    init: bool,
    send_count: usize,
    recv_count: usize,
    world: &'w SimpleCommunicator,
}

impl<'w, I> Iterator for DistIter<'w, I>
where
    I: Iterator,
    I::Item: Task,
{
    type Item = <I::Item as Task>::OUT;

    fn next(&mut self) -> Option<Self::Item> {
        if !self.init {
            for dest in 1..self.world.size() {
                if let Some(task) = self.inner.next() {
                    task.send(self.world.process_at_rank(dest));
                    self.send_count += 1;
                }
            }
        }
        if self.recv_count < self.send_count {
            let tag = <I::Item as Task>::TAG;
            let (data, status) = self.world.any_process().receive_with_tag(tag);
            self.recv_count += 1;

            if let Some(task) = self.inner.next() {
                task.send(self.world.process_at_rank(status.source_rank()));
                self.send_count += 1;
            }
            Some(data)
        } else {
            None
        }
    }
}

pub trait MyIterExt {
    fn into_dist_iter(self, world: &SimpleCommunicator) -> DistIter<Self>
    where
        Self: Iterator + Sized,
        Self::Item: Task;
}

impl<I> MyIterExt for I
where
    I: Iterator,
    I::Item: Task,
{
    fn into_dist_iter(self, world: &SimpleCommunicator) -> DistIter<Self> {
        DistIter {
            inner: self,
            init: false,
            send_count: 0,
            recv_count: 0,
            world,
        }
    }
}
