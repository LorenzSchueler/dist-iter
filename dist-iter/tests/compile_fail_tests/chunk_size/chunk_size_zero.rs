use dist_iter::{map_task, DistIterator};

fn main() {
    let _ = []
        .into_iter()
        .dist_map(map_task!(CHUNK_SIZE = 0, |x: i32| -> i32 { x * x }));
}
