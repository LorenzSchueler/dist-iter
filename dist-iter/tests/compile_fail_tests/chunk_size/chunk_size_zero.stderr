error[E0080]: evaluation of `dist_iter::UninitBuffer::<i32, 0>::VALID` failed
 --> src/uninit_buffer.rs
  |
  |     const VALID: () = assert!(N > 0, "CHUNK_SIZE must be greater than 0");
  |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'CHUNK_SIZE must be greater than 0', $DIR/src/uninit_buffer.rs:19:23
  |
  = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn dist_iter::UninitBuffer::<i32, 0>::new`
  --> src/iter/chunk_distributor.rs
   |
   |             buf: UninitBuffer::new(),
   |                  ^^^^^^^^^^^^^^^^^^^
